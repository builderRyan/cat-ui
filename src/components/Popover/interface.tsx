import * as React from 'react'

export interface genChildProps {
  trigger: any;
  content: any;
}
