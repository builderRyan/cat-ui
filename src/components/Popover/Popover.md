Example

```jsx
import Button from '../Button';

class Example extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      visible1: false
    }
    this.onVisibleChange2 = this.onVisibleChange2.bind(this)
    this.showPopover1 = this.showPopover1.bind(this)
    this.showPopover2 = this.showPopover2.bind(this)
    this.closePopover2 = this.closePopover2.bind(this)
  }

  onVisibleChange2 () {
    const { visible1 } = this.state
    this.setState({ visible1: !visible1 })
  }

  showPopover1 () {
    console.log('showPopover1')
  }

  showPopover2 () {
    this.setState({ visible1: true })
  }

  closePopover2 () {
    this.setState({ visible1: false })
  }

  render () {
    const { visible1 } = this.state

    return (<div className='popover-box'>
      <h3>mode='click'的Popover</h3>
      <Popover mode='click'>
        <Popover.Trigger>
          <Button onClick={this.showPopover1}>点击打开</Button>
        </Popover.Trigger>
        <Popover.Content>
          <div className='pop-content'>
            <p>通用的触发式弹层组件, 可以自定义定位算法、触发方式以及弹层显示方式。</p>
            <p>向下弹出的弹层组件</p>
          </div>
        </Popover.Content>
      </Popover>

       <h3>mode='hover'的Popover</h3>
      <Popover mode='hover'>
        <Popover.Trigger>
          <Button onClick={this.showPopover1}>点击打开</Button>
        </Popover.Trigger>
        <Popover.Content>
          <div className='pop-content'>
            <p>通用的触发式弹层组件, 可以自定义定位算法、触发方式以及弹层显示方式。</p>
            <p>向下弹出的弹层组件</p>
          </div>
        </Popover.Content>
      </Popover>

      <h3>外部visible1的Popover</h3>
      <Popover 
        visible={visible1}
        onVisibleChange={this.onVisibleChange2}>
        <Popover.Trigger>
          <Button onClick={this.showPopover2}>点击打开</Button>
          <Button onClick={this.closePopover2}>点击关闭</Button>
        </Popover.Trigger>
        <Popover.Content>
          <div className='pop-content'>
            <p>通用的触发式弹层组件, 可以自定义定位算法、触发方式以及弹层显示方式。</p>
            <p>向下弹出的弹层组件</p>
          </div>
        </Popover.Content>
      </Popover>

      <style>
        {`
          .popover-box button + button {
            margin-left: 20px;
          }
        `}
      </style>
    </div>
    )
  }
}

<Example />
```
