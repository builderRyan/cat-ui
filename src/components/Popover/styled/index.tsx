import styled, { keyframes } from 'styled-components'
import { primaryBtnColor } from '../../../styles/theme'
import { pageTurning } from '../../../styles/animation'

const StyledPopover = styled.div`
  display: inline-block;
`;

const StyledPopoverTrigger = styled.div``;

const StyledPopoverContent = styled.div`
  position: relative;
  border: 1px solid #e5e5e5;
  /* padding: 10px; */
  border-radius: 4px;
  font-size: 14px;
`;

const StyledPopBox = styled.div`
  position: relative;
  border-radius: 4px;
  background: #fff;
  font-size: 14px;
  transform-origin: left center;
  animation: .15s ${pageTurning} ease-in;
`

export { StyledPopover, StyledPopoverTrigger, StyledPopoverContent, StyledPopBox }
