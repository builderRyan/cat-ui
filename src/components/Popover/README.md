# Popover 弹出层组件

- 弹出层组件，适用于 `一个容器`（content） 相对于 `另外一个容器`（trigger） `定位` 的弹出层组件，如 `Tooltip`, `Select` 组件

## Example

```jsx
import { positionNameConvert } from '../Popover/utils';

const App = () => {
  // 渲染内容
  const renderTooltipContent = (visible: boolean) => {
    return (
      <CSSTransition
        timeout={300}
        in={visible}
        unmountOnExit
        mountOnEnter
        appear
      >
        popover content
      </CSSTransition>
    );
  }

  const { triggerPosition, contentPosition } = positionNameConvert(position);

  return (
    <Popover
      triggerPosition={triggerPosition}
      contentPosition={contentPosition}
      containerSelector='body'
    >
      <Popover.Trigger>popover trigger</Popover.Trigger>
      <Popover.Content>{renderTooltipContent}</Popover.Content>
    </Popover>
  )
}

```

- 在 `catui` 中，定义了 `12` 种 `content` 相对于 `trigger` 的位置（position），为：`top, topLeft, topRight, right, rightTop, rightBottom, bottom, bottomLeft, bottomRight, left, leftTop, leftBottom`。可以通过使用 `positionNameCover(position)` 获取 `Popover` 组件所需要的 `triggerPosition` 和 `contentPosition` 属性（具体使用可参考 `Tooltip` 组件源码）



## Popover

- 用于组合 `Popover.Trigger` 和 `Popover.Content` 组件

|      属性       |                   类型                   |        默认值        |           描述            | 是否必选 |
| :-------------: | :--------------------------------------: | :------------------: | :-----------------------: | :------: |
|    selector     |                 `string`                 |        `body`        | 弹层所在容器的 css 选择器 |    -     |
|     visible     |                `boolean`                 |          -           |       弹层是否可见        |    -     |
|      mode       |                `ModeType`                |       `click`        |         触发类型          |    -     |
| triggerPosition | `[HorizontalPosition, VerticalPosition]` | `['left', 'bottom']` |       trigger 方位        |    是    |
| contentPosition | `[HorizontalPosition, VerticalPosition]` |  `['left', 'top']`   |       content 方位        |    是    |
|     offsetX     |                 `number`                 |         `0`          |    水平方向上的偏移量     |    -     |
|     offsetY     |                 `number`                 |         `0`          |    垂直方向上的偏移量     |    -     |
|    onChange     |       `(visible: boolean) => void`       |          -           |   visible 变化时的回调    |    -     |

```typescript
export type ModeType = Array<'click' | 'hover'> | 'click' | 'hover';

export type HorizontalPosition = 'left' | 'center' | 'right';

export type VerticalPosition = 'top' | 'center' | 'bottom';
```

## Popover.Trigger

- 触发弹出层的容器


|   属性   |         类型         | 默认值 |       描述       | 是否必选 |
| :------: | :------------------: | :----: | :--------------: | :------: |
| children | `React.ReactElement` |   -    | 触发弹出层的容器 |    是    |

## Popover.Content

- 弹出层

|   属性   |                  类型                   | 默认值 |                           描述                           | 是否必选 |
| :------: | :-------------------------------------: | :----: | :------------------------------------------------------: | :------: |
| children | `(visible: boolean) => React.ReactNode` |   -    | 返回弹出层的回调（参数 visible 可以用来显示/隐藏弹出层） |    是    |

- 可以结合 `CSSTransition` 实现显示/隐藏的动画
