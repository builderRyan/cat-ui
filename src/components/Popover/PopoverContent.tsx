import * as React from 'react'
import { Component } from 'react'
import classNames from 'classnames'
import * as throttle from 'lodash/throttle'
import { StyledPopoverContent, StyledPopBox } from './styled'
import Portal from '../Portal'
// import WindowEventHandler from '../../utils/WindowEventHandler'
// import WindowResizeHandler from '../../utils/WindowResizeHandler'

export interface PopoverContentProps {
  /** 类名 */
  className?: string;
  /** 样式 */
  style?: object;
  /** 弹层是否可见 */
  visible?: boolean;
  /** triggerDOM */
  triggerDOM?: any;
  /** 触发类型 */
  mode?: 'click' | 'hover';
  /** 定位的方向 */
  position?: 'bottomLeft' | 'topCenter';
  /** toggleVisible */
  toggleVisible?: any;
}

const defaultPostion = {
  position: 'absolute',
  top: 0,
  left: 0,
}

/**
 * PopoverContent
 */
class PopoverContent extends Component<PopoverContentProps, any> {
  private contentRef: any;
  constructor (props: PopoverContentProps) {
    super(props)
    this.state = {
      positionStyle: defaultPostion
    }
    this.contentRef = React.createRef()
  }

  componentDidMount() {
    const { visible } = this.props
    if (visible) {
      this.adjustPosition()
    }
    window.addEventListener('scroll', this.onWindowScroll)
    window.addEventListener('resize', this.onWindowResize)
  }

  componentDidUpdate(prevProps: PopoverContentProps) {
    if (this.props.visible && prevProps.visible !== this.props.visible) {
      // reset position mark
      // this.positionReady = false;
      this.adjustPosition()
    }
  }

  componentWillUnmount () {
    window.removeEventListener('scroll', this.onWindowScroll)
    window.removeEventListener('resize', this.onWindowResize)
  }

  adjustPosition = () => {
    const { visible, triggerDOM, position } = this.props
    if (!visible || !triggerDOM) { return }
    const { positionStyle } =  this.state
    const contentDOM = this.contentRef.current
    const triggerRect = triggerDOM.getBoundingClientRect()
    const contentRect = contentDOM.getBoundingClientRect()

    let newPositionStyle

    switch (position) {
      case 'topCenter':
        newPositionStyle =  {
          ...positionStyle,
          left: triggerRect.left + (triggerRect.width / 2) - (contentRect.width / 2),
          top: window.pageYOffset + triggerRect.top - contentRect.height - 5
        }
        break
      case 'bottomLeft':
      default:
        newPositionStyle =  {
          ...positionStyle,
          left: triggerRect.left,
          top: window.pageYOffset + triggerRect.top + triggerRect.height
        }
    }

    this.setState({ positionStyle: newPositionStyle })
  }

  onWindowResize = throttle((evt: any) => {
    if (this.props.visible) {
      this.adjustPosition()
    }
  }, 16)

  onWindowScroll = throttle(this.adjustPosition, 16)

  handleMouseEnter = () => {
    const { mode, toggleVisible } = this.props
    if (mode === 'hover') {
      toggleVisible(true)
    }
  }

  handleMouseLeave = () => {
    const { mode, toggleVisible } = this.props
    if (mode === 'hover') {
      toggleVisible(false)
    }
  }

  render() {
    const { positionStyle } = this.state
    const { className, style, visible, triggerDOM, children } = this.props
    const classes = classNames('hmly-popover', className)
    // console.log('PopoverContent:render', visible, triggerDOM)
    // console.log('PopoverContent:visible', visible)

    return (
      <Portal visible={visible}>
        <StyledPopBox
          ref={this.contentRef}
          style={positionStyle}
          onMouseEnter={this.handleMouseEnter}
          onMouseLeave={this.handleMouseLeave}>
          <StyledPopoverContent
            className={classes}
            style={style}>
            {children}
          </StyledPopoverContent>
        </StyledPopBox>
      </Portal>
    )
  }
}

export default PopoverContent
