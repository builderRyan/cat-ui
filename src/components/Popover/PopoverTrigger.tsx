import * as React from 'react'
import { Component } from 'react'
import classNames from 'classnames'
import { StyledPopoverTrigger } from './styled'

export interface PopoverTriggerProps {
  /** 类名 */
  className?: string;
  /** 样式 */
  style?: object;
  /** 弹层是否可见 */
  visible?: boolean;
}

/**
 * PopoverTrigger
 */
class PopoverTrigger extends Component<PopoverTriggerProps, any> {

  render() {
    const { className, style, children } = this.props
    const classes = classNames('hmly-popover', className)

    return (
      <StyledPopoverTrigger
        className={classes}
        style={style}>
        {children}
      </StyledPopoverTrigger>
    )
  }
}

export default PopoverTrigger
