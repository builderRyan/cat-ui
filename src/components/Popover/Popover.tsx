import * as React from 'react'
import { Component, Children, cloneElement } from 'react'
import classNames from 'classnames'
import { StyledPopover } from './styled'
import { genChildProps } from './interface'
import PopoverTrigger from './PopoverTrigger'
import PopoverContent from './PopoverContent'
import kindOf from '../../utils/kindOf'

export interface PopoverProps {
  /** 类名 */
  className?: string;
  /** 样式 */
  style?: object;
  /** 弹层是否可见 */
  visible?: boolean;
  /** 触发类型 */
  mode: 'click' | 'hover';
}

/**
 * 弹层
 */
class Popover extends Component<PopoverProps, any> {
  private popoverRef: any;
  static Trigger: typeof PopoverTrigger;
  static Content: typeof PopoverContent;
  static defaultProps = {
    mode: 'click'
  }

  constructor (props: PopoverProps) {
    super(props)
    this.state = {
      visible: false
    }
    this.popoverRef = React.createRef()
  }

  static getDerivedStateFromProps (nextProps: PopoverProps) {
    if ('visible' in nextProps) {
      return {
        visible: nextProps.visible
      }
    } else {
      return null
    }
  }

  componentDidMount () {
    window.addEventListener('click', this.removePopover)
  }

  componentWillUnmount () {
    window.removeEventListener('click', this.removePopover)
  }

  open = () => {

  }

  close = () => {

  }

  toggleVisible = (value: boolean) => {
    this.setState({ visible: value })
  }

  removePopover = (e: any) => {
    const { visible } = this.state
    const node = e.target
    const triggerDOM = this.popoverRef.current

    // // 如果点击的节点不在popup中或者有clickRemove属性
    if (visible && triggerDOM.contains(node) === false) {
      // this.togglePopup(false)
      console.log('removePopover')
      this.setState({ visible: false })
    }
  }

  genTriggerContent = () => {
    const { children } = this.props
    const childArray = Children.toArray(children)
    if (childArray.length !== 2) {
      throw new Error('There must be one and only one trigger and content in Popover')
    }
    const { trigger, content } = childArray.reduce <genChildProps> (
      (result, child: React.ReactElement<any>) => {
        const type = child.type
        if (kindOf(type, PopoverTrigger)) {
          result.trigger = child
        } else if (kindOf(type, PopoverContent)) {
          result.content = child
        }

        return result
      },
      { trigger: null, content: null }
    )
    return { trigger, content }
  }

  renderTrigger = (trigger: any) => {
    const { mode } =  this.props
    const { visible } = this.state
    return React.cloneElement(trigger, {
      mode: mode,
      open: this.open,
      close: this.close,
      toggleVisible: this.toggleVisible
    })
  }

  renderContent = (content: any) => {
    const { mode } = this.props
    const { visible } = this.state
    return React.cloneElement(content, {
      visible: visible,
      mode: mode,
      open: this.open,
      close: this.close,
      triggerDOM: this.popoverRef.current
    })
  }

  render() {
    const { className, style, children } = this.props
    const { visible } = this.state
    const classes = classNames('hmly-popover', className)
    const { trigger, content } = this.genTriggerContent()
    // const visible = this.getVisible()

    console.log('Popover:render', visible)

    return (
      <StyledPopover
        ref={this.popoverRef}
        className={classes}
        style={style}>
        {this.renderTrigger(trigger)}
        {this.renderContent(content)}
      </StyledPopover>
    )
  }
}

export default Popover
