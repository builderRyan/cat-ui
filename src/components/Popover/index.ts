import Popover from './Popover'
import PopoverTrigger from './PopoverTrigger'
import PopoverContent from './PopoverContent'

Popover.Trigger = PopoverTrigger
Popover.Content = PopoverContent

export default Popover
