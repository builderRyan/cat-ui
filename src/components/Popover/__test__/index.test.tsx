import * as React from 'react'
import * as Enzyme from 'enzyme'
import { render, mount } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'
import Popover from '../index'

Enzyme.configure({ adapter: new Adapter() })

describe('Popover', () => {
  test('renders correctly', () => {
    const wrapper = render(
      <Popover mode='click'>
        <Popover.Trigger>
          <span>点击打开</span>
        </Popover.Trigger>
        <Popover.Content>
          <div className='pop-content'>
            <p>通用的触发式弹层组件, 可以自定义定位算法、触发方式以及弹层显示方式。</p>
            <p>向下弹出的弹层组件</p>
          </div>
        </Popover.Content>
      </Popover>
    )
    expect(wrapper).toMatchSnapshot()
  })

})
