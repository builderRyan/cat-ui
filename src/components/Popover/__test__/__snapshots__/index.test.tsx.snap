// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover renders correctly 1`] = `
<Popover
  contentPosition={
    Array [
      "left",
      "top",
    ]
  }
  mode="click"
  offsetX={0}
  offsetY={0}
  onChange={[Function]}
  selector="body"
  triggerPosition={
    Array [
      "left",
      "bottom",
    ]
  }
>
  <PopoverTrigger
    disabled={false}
    key=".0"
    mode="click"
    toggleVisible={[Function]}
  >
    <span
      onClick={[Function]}
    >
      popover trigger
    </span>
  </PopoverTrigger>
  <PopoverContent
    contentPosition={
      Array [
        "left",
        "top",
      ]
    }
    key=".1"
    mode="click"
    offsetX={0}
    offsetY={0}
    selector="body"
    toggleVisible={[Function]}
    triggerDOM={
      <span>
        popover trigger
      </span>
    }
    triggerPosition={
      Array [
        "left",
        "bottom",
      ]
    }
    visible={false}
  >
    <PurePortal
      selector="body"
    >
      <Portal
        containerInfo={
          <body>
            <div
              class="cat-popover__content"
              style="position: absolute; top: 0px; left: 0px;"
            >
              <div>
                popover content
              </div>
            </div>
          </body>
        }
      >
        <div
          className="cat-popover__content"
          onClick={[Function]}
          style={
            Object {
              "left": 0,
              "position": "absolute",
              "top": 0,
            }
          }
        >
          <div>
            popover content
          </div>
        </div>
      </Portal>
    </PurePortal>
  </PopoverContent>
</Popover>
`;
